File: 62-unique-paths/unique-paths.cpp
class Solution {
public:
    int uniquePaths(int m, int n) {
        vector<vector<int>> dp(2,vector<int>(n));
        for(int i=0;i<n;++i) dp[0][i]=1;
        for(int i=0;i<2;++i) dp[i][0]=1;
        for(int r=1;r<m;++r){
            for(int c=1;c<n;++c){
                dp[r%2][c]=dp[r%2][c-1]+dp[1-r%2][c];
            }
        }
        return dp[(m-1)%2][n-1];
    }
};

File: 62-unique-paths/unique-paths.java
class Solution {
    private int uniquePathsUtil(int i, int j, int[][] dp){
        if(i==0 || j==0) return 1;
        if(dp[i][j]==-1){
            dp[i][j]= uniquePathsUtil(i-1, j, dp) + uniquePathsUtil(i, j-1, dp);
        }
        return dp[i][j];
    }
    public int uniquePaths(int m, int n) {
        int[][] dp = new int[m][n];
        for(int i=0;i<m;i++){
            Arrays.fill(dp[i], -1);
        }
        return uniquePathsUtil(m-1, n-1, dp);
    }
}